### DISCUSSION TITLE
Resolved

### DISCUSSION THREAD

**Discussion by kendfss** [nushell/nushell#6233] [Q&A]
I can't tell if the following is expected or not but it seems strange

I'm writing a shortcut for [the github cli](https://github.com/cli/cli).
But the syntax fails when I wrap it in a function that i'm trying to source into my env file.

here's the basic syntax for the api request (it checks for your starred repos)
```sh
let tst = (gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0})
```
remove `--paginate` if it's too slow.

here are the processing pipelines that work in the REPL, but not in a function, followed by the error paths.
```ruby
$tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
```
[nu::parser::parse_mismatch](https://docs.rs/nu-parser/0.65.0/nu_parser/enum.ParseError.html#variant.Expected)

```ruby
$tst | into df | rename column1 user | rename column2 repo | rename column3 updated | rename column4 first_push | into nu
```
[nu::parser::extra_positional](https://docs.rs/nu-parser/0.65.0/nu_parser/enum.ParseError.html#variant.ExtraPositional)

here's the function syntax
```ruby
def starred [count:int = 30] {
    let tst = ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count)
    # $tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
    # $tst | into df | rename column1 user | rename column2 repo | rename column3 updated | rename column4 first_push | into nu
}
```

----

using `0.65.0`

There were some other variations too, notably using the `-c` flag. I think they also had one of the aforementioned errors.
Is there a way around it/better approach?

**Comment by fdncred** [comment]
You can change your second line of your custom command to this.
```
$tst | rename user repo updated first_push
```
or just add the `| rename user repo updated first_push` to the end of the first line in your custom command.

**Reply by kendfss** [reply]
those throw [nu::parser::extra_positional](https://docs.rs/nu-parser/0.65.0/nu_parser/enum.ParseError.html#variant.ExtraPositional).

**Reply by fdncred** [reply]
What I typed works fine on the latest main. Maybe you're on an old version?
![image](https://user-images.githubusercontent.com/343840/183075511-ed94697b-e575-4645-8210-f71a8a77d22a.png)

**Reply by kendfss** [reply]
What version are you on?
I was on `0.65.0`, then I cloned and built from source and am on `0.66.0`. Still no luck, same errors.

**Reply by kendfss** [reply]
Also, for the record, I agree that a lot of what I posted looks superfluous but until I changed it into a dataframe I couldn't rename columns at all.

**Reply by fdncred** [reply]
Here's my version information. It looked to me like you just had the wrong syntax for renaming regular nushell columns.

|key|value|
|-|-|
|version|0.66.3|
|branch|main|
|commit_hash|78a5067434ad85fa54e8d7971e0f17017bd423b2|
|build_os|macos-aarch64|
|build_target|aarch64-apple-darwin|
|rust_version|rustc 1.62.1 (e092d0b6b 2022-07-16)|
|rust_channel|stable-aarch64-apple-darwin|
|cargo_version|cargo 1.62.1 (a748cf5a3 2022-06-08)|
|pkg_version|0.66.3|
|build_time|2022-08-04 16:33:54 -05:00|
|build_rust_channel|release|
|features|database, dataframe, default, trash, which, zip|
|installed_plugins|from parquet, gstat, inc, query, query json, query web, query xml|

**Reply by kendfss** [reply]
But it works in the REPL. Btw, how do you get the detailed version info? by hand or is there a pipeline for it?

**Reply by fdncred** [reply]
`version | transpose key value` and if I want to post it on GitHub I add a `| to md`

**Reply by kendfss** [reply]
Cheers. Never knew there was a version command, was just using the flag from another shell.
I checked out and built 78a5067 and it also has the same issues.

The commit hash can be found by `git rev-parse --short HEAD`, right? If so, after updating again, mine is c2f4969d4. Here's what I'm using (and still having the same errors).

|key|value|
|-|-|
|version|0.66.3|
|branch|main|
|commit_hash||
|build_os|linux-x86_64|
|build_target|x86_64-unknown-linux-gnu|
|rust_version|rustc 1.62.1 (e092d0b6b 2022-07-16)|
|cargo_version|cargo 1.62.1 (a748cf5a3 2022-06-08)|
|pkg_version|0.66.3|
|build_time|2022-08-06 01:38:56 +01:00|
|build_rust_channel|release|
|features|database, dataframe, default, trash, which, zip|
|installed_plugins||


And here's a, admittedly dubiously helpful, neofetch.

|key|value|
|-|-|
|OS|Linux Mint 20.3 x86_64|
|Host|MacBookAir7,2 1.0|
|Kernel|5.4.0-122-generic|
|Uptime|4 days, 16 hours, 32 mins|
|Packages|2753 (dpkg)|
|Shell|nu 0.66.2|
|Resolution|1440x900|
|DE|Cinnamon|
|WM|Mutter (Muffin)|
|WM theme|Mint-Y-Dark-Grey (Mint-Y)|
|Theme|Mint-Y-Dark-Grey [GTK2/3]|
|Icons|Mint-Y-Dark-Grey [GTK2/3]|
|Terminal|gnome-terminal|
|CPU|Intel i5-5350U (4) @ 2.900GHz|
|GPU|Intel HD Graphics 6000|
|Memory|2894MiB / 7860MiB|


Should my branch and commit hash automatically be included in the `version` output or did you add those manually in your paste?

---


All that said, the nature of the issue seems different. Now when I use:
```nu
let tst = ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count)

$tst | rename user repo updated first_push
```
In the repl I get the extra-positional error, pointing at `repo`.
Which I think is what lead me to that convoluted pipeline in the first place. For some reason, the `rename` command only expects one positional argument. Reinserting that old pipeline (`$tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu`) fixes the REPL but still doesn't work when I source it. I get `expected non-[] value` pointing at the first array, of numbered columns.
Would appreciate if you could recommend a good starting point for further trouble shooting.

**Reply by fdncred** [reply]
> or did you add those manually in your paste?

it should be there. i didn't add it manually.

I'm not having problems. Not sure what the issue is, perhaps the data is coming back odd?
![image](https://user-images.githubusercontent.com/343840/184405487-d9c7b38b-94ca-4cd7-99a9-833e79dd237b.png)

for troubleshooting i'd start looking at the data output from gh to see if something looks odd. then i'd do the split row each part and look at that to see if there are issues.

I'd also look into putting the `first $count` part into the gh api side because you're always returning everything now and then just taking the first x items. It would be faster if you tell gh to get the first x and then remove that part of the nushell pipeline at the end

**Reply by kendfss** [reply]
I don't think you can break out of iteration using go templates and there's no cli flag for choosing the number of results ahead of time. You either get ~30, or the whole thing.

I can do what you're doing on the REPL, and can even put it all on one line, but I cannot get it to work when I try to use the `rename` command within the scope of a function definition. I've checked for invisible characters but can't see anything unusual.

here's my tree
```bash
$ tree
.
├── config.nu
├── env.nu
├── functions.nu
└── login.nu
```

at the bottom of `env.nu` I have
```bash
source "/home/kendfss/.config/nushell/functions.nu"
```
in which I have
```ruby
def starred [count:int = 30] {
    # let tst = ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count)
    ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count)
     # | rename user repo updated first_push
    # $tst
    # $tst | rename user #repo updated first_push
    # $tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
    # $tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
    # $tst | into df | rename column1 user | rename column2 repo | rename column3 updated | rename column4 first_push | into nu
}
```

but if I edit the definition to rename the columns I get the errors.

**Reply by fdncred** [reply]
I just did it like tihs and it works fine for me. I wonder why yours doesn't work?
```ruby
def starred [count:int = 30] {
    # let tst = ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count)
    ((gh api user/starred --paginate --template '{{range .}}{{.owner.login}}{{"*"}}{{.name}}{{"*"}}{{timeago .updated_at}}{{"*"}}{{timeago .pushed_at}}{{"\n"}}{{end}}' | split row "\n" | each {|it| ($it | split column "*").0}) | first $count) | rename user repo updated first_push
    # $tst
    # $tst | rename user #repo updated first_push
    # $tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
    # $tst | into df | rename [column1 column2 column3 column4] [user repo updated first_push] | into nu
    # $tst | into df | rename column1 user | rename column2 repo | rename column3 updated | rename column4 first_push | into nu
}
```
![image](https://user-images.githubusercontent.com/343840/184654809-6ed46d5c-efe1-41ed-b795-9b613f0ca442.png)

**Comment by kendfss** [comment]
It says because of [nu::parser::extra_positional](https://docs.rs/nu-parser/0.66.2/nu_parser/enum.ParseError.html#variant.ExtraPositional) but I do not understand why.

![myimage_000](https://user-images.githubusercontent.com/73350225/185837623-a5f37edb-8f58-4a29-94ef-10c2eeb8da1c.png)

**Reply by fdncred** [reply]
The command I pasted above works. The only conclusion that I can come up with is that your data is sometimes returning without all the columns. So, if you had a table without the repo column for one of your results, I think you'd get this same error.

**Reply by kendfss** [reply]
looool, found the bug!
![image](https://user-images.githubusercontent.com/73350225/187186187-af5f62e5-3547-48ec-b327-6f35e1e446e7.png)

**Reply by kendfss** [reply]
would it be a horrible time to ask if your team has any plans for allowing directory changes within a function definition?

**Reply by fdncred** [reply]
Directory changes are allowed within custom commands. The custom command has to start with `def-env` though instead of `def`.
